test/progress.js

```diff
@@ -0,0 +1,197 @@
+
+		intoStream(file)
+			.pipe(new SlowStream({maxWriteInterval: 50}))
+			.pipe(res);
+	});
+
+	s.on('/download/no-total', (req, res) => {
+		res.write('hello');
+		res.end();
+	});
+
+	s.on('/upload', (req, res) => {
+		req
+			.pipe(new SlowStream({maxWriteInterval: 100}))
+			.on('end', () => res.end());
+	});
+
+	await s.listen(s.port);
+});
+
+test('download progress', async t => {
+	const events = [];
+
+	const res = await got(`${s.url}/download`, {encoding: null})
+		.on('downloadProgress', e => events.push(e));
+
+	checkEvents(t, events, res.body.length);
+});
+
+test('download progress - missing total size', async t => {
+	const events = [];
+
+	await got(`${s.url}/download/no-total`)
+		.on('downloadProgress', e => events.push(e));
+
+	checkEvents(t, events);
+});
+
+test.cb('download progress - stream', t => {
+	const events = [];
+
+	got.stream(`${s.url}/download`, {encoding: null})
+		.on('downloadProgress', e => events.push(e))
+		.on('data', () => {})
+		.on('end', () => {
+			checkEvents(t, events, file.length);
+			t.end();
+		});
+});
+
+test('upload progress - file', async t => {
+	const events = [];
+
+	await got.post(`${s.url}/upload`, {body: file})
+		.on('uploadProgress', e => events.push(e));
+
+	checkEvents(t, events, file.length);
+});
+
+test('upload progress - file stream', async t => {
+	const path = tempfile();
+	fs.writeFileSync(path, file);
+
+	const events = [];
+
+	await got.post(`${s.url}/upload`, {body: fs.createReadStream(path)})
+		.on('uploadProgress', e => events.push(e));
+
+	checkEvents(t, events, file.length);
+});
+
+test('upload progress - form data', async t => {
+	const events = [];
+
+	const body = new FormData();
+	body.append('key', 'value');
+	body.append('file', file);
+
+	const size = await pify(body.getLength.bind(body))();
+
+	await got.post(`${s.url}/upload`, {body})
+		.on('uploadProgress', e => events.push(e));
+
+	checkEvents(t, events, size);
+});
+
+test('upload progress - json', async t => {
+	const body = JSON.stringify({key: 'value'});
+	const size = Buffer.byteLength(body);
+	const events = [];
+
+	await got.post(`${s.url}/upload`, {body})
+		.on('uploadProgress', e => events.push(e));
+
+	checkEvents(t, events, size);
+});
+
+test.cb('upload progress - stream with known body size', t => {
+	const events = [];
+	const options = {
+		headers: {'content-length': file.length}
+	};
+
+	const req = got.stream.post(`${s.url}/upload`, options)
+		.on('uploadProgress', e => events.push(e));
+
+	intoStream(file)
+		.pipe(req)
+		.on('data', () => {})
+		.on('end', () => {
+			checkEvents(t, events, file.length);
+			t.end();
+		});
+});
+
+test.cb('upload progress - stream with unknown body size', t => {
+	const events = [];
+
+	const req = got.stream.post(`${s.url}/upload`)
+		.on('uploadProgress', e => events.push(e));
+
+	intoStream(file)
+		.pipe(req)
+		.on('data', () => {})
+		.on('end', () => {
+			checkEvents(t, events);
+			t.end();
+		});
```

Comment:
You could use https://github.com/sindresorhus/get-stream here to be able to use async/await instead of `test.cb`.